smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/provider/UserTuneProvider.java

```diff
@@ -0,0 +1,115 @@
+    public static final UserTuneProvider INSTANCE = new UserTuneProvider();
+
+    @Override
+    public UserTuneElement parse(XmlPullParser parser, int initialDepth, XmlEnvironment xmlEnvironment)
+            throws XmlPullParserException, IOException, SmackParsingException {
+
+        UserTuneElement.Builder builder = UserTuneElement.getBuilder();
+        outerloop: while (true) {
+            XmlPullParser.Event tag = parser.next();
+            String name = parser.getName();
+            String namespace = parser.getNamespace();
+
+            switch (tag) {
+            case START_ELEMENT:
+                if (!UserTuneElement.NAMESPACE.equals(namespace)) {
+                    LOGGER.log(Level.FINE, "Foreign namespace " + namespace + " detected.");
+                    continue outerloop;
+                }
+                try {
+                    while (true) {
+                        switch (name) {
+                        case "artist":
+                            builder.setArtist(parser.nextText());
+                            break;
+                        case "length":
+                            builder.setLength(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "rating":
+                            builder.setRating(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "source":
+                            builder.setSource(parser.nextText());
+                            break;
+                        case "title":
+                            builder.setTitle(parser.nextText());
+                            break;
+                        case "track":
+                            builder.setTrack(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "uri":
+                            URI uri = new URI(parser.nextText());
+                            builder.setUri(uri);
+                            break;
+                        case "tune":
+                            if (UserTuneElement.ELEMENT.equals(name) && (tag == Event.END_ELEMENT)) {
+                                return builder.build();
+                            }
+                        }
+                        tag = parser.next();
+                        name = parser.getName();
+                        if (isNameInvalid(name)) {
+                            throw new XmlPullParserException("Invalid elements inside <tune/>");
+                        }
+                    }
+                } catch (IllegalArgumentException e) {
+                    throw new XmlPullParserException("Unknown tune value: " + name + " encountered.");
+                } catch (URISyntaxException e) {
+                    throw new IllegalArgumentException("Invalid URI");
+                }
+
+            case END_ELEMENT:
+                if (UserTuneElement.ELEMENT.equals(name)) {
```

Comment:
This should take (only) `depth` into account, which ensures we actually hit the correct closing tag (and not a closing tag of a different depth which just happens to have the same name).